`yesod-persistent-1.6.0.7` fails to build with `persistent-2.14.0.0`

<!---

### Bug Reports

Reporting a bug? If relevant, we recommend including:

* Your OS name and version
* The versions of tools you're using (e.g. `stack`, `yesod` `ghc`).
* The versions of dependencies you're using

For your convenience, we recommend pasting this script into bash and uploading the output [as a gist](https://gist.github.com/).

```
command -v sw_vers && sw_vers # OS X only
command -v uname && uname -a # Kernel version
command -v stack && stack --version
command -v stack && stack ghc -- --version
command -v stack && stack ls dependencies
command -v yesod && yesod version
```

* Also, is there anything custom or unusual about your setup? i.e. new or prerelease versions of GHC, stack, etc.

* Finally, if possible, please reproduce the error in a small script, or if necessary create a new Github repo with the smallest possible reproducing case. [Stack's scripting support](https://docs.haskellstack.org/en/stable/GUIDE/#script-interpreter) might be useful for creating your reproduction example.

### Support

Please direct support questions to [Stack Overflow](https://stackoverflow.com/questions/tagged/yesod+haskell) or the [Yesod Google Group](https://groups.google.com/forum/#!forum/yesodweb). If you don't get a response there, or you suspect there may be a bug in Yesod causing your problem, you're welcome to ask here.
-->

`yesod-persistent-1.6.0.7` fails to build with `persistent-2.14.0.0`:

```
$ cabal repl -b yesod-persistent --constraint="persistent==2.14.0.0" -w ghc-9.0.2
Resolving dependencies...
Build profile: -w ghc-9.0.2 -O1
In order, the following will be built (use -v for more details):
 - yesod-persistent-1.6.0.7 (lib) (requires build)
 - fake-package-0 (lib) (first run)
Starting     yesod-persistent-1.6.0.7 (lib)
Building     yesod-persistent-1.6.0.7 (lib)

Failed to build yesod-persistent-1.6.0.7.
Build log (
/home/ryanglscott/.cabal/logs/ghc-9.0.2/yesod-persistent-1.6.0.7-a3bdbcb14a59518d7dd131eecf33a57d2ec29c106c04f7e7aa49ef8edbea502c.log
):
Configuring library for yesod-persistent-1.6.0.7..
Preprocessing library for yesod-persistent-1.6.0.7..
Building library for yesod-persistent-1.6.0.7..
[1 of 2] Compiling Yesod.Persist.Core ( Yesod/Persist/Core.hs, dist/build/Yesod/Persist/Core.o, dist/build/Yesod/Persist/Core.dyn_o )

Yesod/Persist/Core.hs:212:20: error:
    • Could not deduce (Database.Persist.Class.PersistEntity.SafeToInsert
                          val)
        arising from a use of ‘insert’
      from the context: (MonadIO m, PersistUniqueWrite backend,
                         PersistRecordBackend val backend)
        bound by the type signature for:
                   insert400 :: forall (m :: * -> *) backend val.
                                (MonadIO m, PersistUniqueWrite backend,
                                 PersistRecordBackend val backend) =>
                                val -> ReaderT backend m (Key val)
        at Yesod/Persist/Core.hs:(191,1)-(193,40)
      Possible fix:
        add (Database.Persist.Class.PersistEntity.SafeToInsert
               val) to the context of
          the type signature for:
            insert400 :: forall (m :: * -> *) backend val.
                         (MonadIO m, PersistUniqueWrite backend,
                          PersistRecordBackend val backend) =>
                         val -> ReaderT backend m (Key val)
    • In the expression: insert datum
      In a case alternative: Nothing -> insert datum
      In a stmt of a 'do' block:
        case conflict of
          Just unique
            -> badRequest'
                 $ map (unFieldNameHS . fst)
                     $ toList $ persistUniqueToFieldNames unique
          Nothing -> insert datum
    |
212 |         Nothing -> insert datum
    |                    ^^^^^^^^^^^^
cabal: Failed to build yesod-persistent-1.6.0.7 (which is required by
fake-package-0). See the build log above for details.
```

On the other hand, it compiles if I use `persistent-2.13.3.5` instead.

--- a/yesod-persistent/ChangeLog.md
+++ b/yesod-persistent/ChangeLog.md
@@ -2,7 +2,7 @@
 
 ## 1.6.0.8
 
-* Add support for `persistent-2.14` [#][()
+* Add support for `persistent-2.14` [#1706](https://github.com/yesodweb/yesod/pull/1760)
 
 ## 1.6.0.7
 

--- a/yesod-persistent/ChangeLog.md
+++ b/yesod-persistent/ChangeLog.md
@@ -1,5 +1,9 @@
 # ChangeLog for yesod-persistent
 
+## 1.6.0.8
+
+* Add support for `persistent-2.14` [#][()
+
 ## 1.6.0.7
 
 * Add support for persistent 2.13. [#1723](https://github.com/yesodweb/yesod/pull/1723)

--- a/yesod-persistent/Yesod/Persist/Core.hs
+++ b/yesod-persistent/Yesod/Persist/Core.hs
@@ -37,6 +37,9 @@ import qualified Database.Persist.Sql as SQL
 #if MIN_VERSION_persistent(2,13,0)
 import qualified Database.Persist.SqlBackend.Internal as SQL
 #endif
+#if MIN_VERSION_persistent(2,14,0)
+import Database.Persist.Class.PersistEntity
+#endif
 
 unSqlPersistT :: a -> a
 unSqlPersistT = id
@@ -187,14 +190,21 @@ getBy404 key = do
 -- is violated.
 --
 -- @since 1.4.1
-#if MIN_VERSION_persistent(2,5,0)
-insert400 :: (MonadIO m, PersistUniqueWrite backend, PersistRecordBackend val backend)
-          => val
-          -> ReaderT backend m (Key val)
+#if MIN_VERSION_persistent(2,14,0)
+insert400
+    :: (MonadIO m, PersistUniqueWrite backend, PersistRecordBackend val backend, SafeToInsert val)
+    => val
+    -> ReaderT backend m (Key val)
+#elif MIN_VERSION_persistent(2,5,0)
+insert400
+    :: (MonadIO m, PersistUniqueWrite backend, PersistRecordBackend val backend)
+    => val
+    -> ReaderT backend m (Key val)
 #else
-insert400 :: (MonadIO m, PersistUnique (PersistEntityBackend val), PersistEntity val)
-          => val
-          -> ReaderT (PersistEntityBackend val) m (Key val)
+insert400
+    :: (MonadIO m, PersistUnique (PersistEntityBackend val), PersistEntity val)
+    => val
+    -> ReaderT (PersistEntityBackend val) m (Key val)
 #endif
 insert400 datum = do
     conflict <- checkUnique datum
@@ -214,7 +224,12 @@ insert400 datum = do
 -- | Same as 'insert400', but doesn’t return a key.
 --
 -- @since 1.4.1
-#if MIN_VERSION_persistent(2,5,0)
+#if MIN_VERSION_persistent(2,14,0)
+insert400_ :: (MonadIO m, PersistUniqueWrite backend, PersistRecordBackend val backend, SafeToInsert val)
+           => val
+           -> ReaderT backend m ()
+
+#elif MIN_VERSION_persistent(2,5,0)
 insert400_ :: (MonadIO m, PersistUniqueWrite backend, PersistRecordBackend val backend)
            => val
            -> ReaderT backend m ()

--- a/yesod-persistent/yesod-persistent.cabal
+++ b/yesod-persistent/yesod-persistent.cabal
@@ -1,6 +1,6 @@
 cabal-version:   >= 1.10
 name:            yesod-persistent
-version:         1.6.0.7
+version:         1.6.0.8
 license:         MIT
 license-file:    LICENSE
 author:          Michael Snoyman <michael@snoyman.com>

